### PR TITLE
[Feature] minimum commission rate upgrade

### PR DESCRIPTION
**Pull request by YunSuk-Yeo**
## Summary of changes
This PR is still under consensus and need governance proposal to be reflected on chain.

### What're Changed

* Add upgrade handler to make validators commission to bigger than minimum rate (= initially 5%)
* Implement ante handler to check whether MsgCreateValidator and MsgEditValidator's commission rate is bigger than minimum rate (= initially 5%)
* New parameter is introduced `ante.MinimumCommissionEnforced`.
* New ante store value is introduced `MinimumCommission`.
* New gRPC endpoint `/terra/ante/v2/params` introduced.
* New gRPC endpoint `/terra/ante/v2/minimum_commission` introduced.
* New cmd `terrad query ante params` introduced.
* New cmd `terrad query ante minimum-commission` introduced.
* New proposal `/terra.ante.v2.MinimumCommissionUpdateProposal` added
   - This proposal will check and update validator commissions, whose commission rate little than the given proposal value.
* New proposal creation cmd `terrad tx gov submit-proposal minimum-commission-update`
```json
{
  ...
  "content": {
    "@type": "/terra.ante.v2.MinimumCommissionUpdateProposal",
    "title": "minimum commission update",
    "description": "minimum commission update desc",
    "minimum_commission": "0.030000000000000000"
  },
  ...
}
``` 

Default `ante.Params`
```go
// Ante params default values
const (
	// DefaultMinimumCommissionEnforced minimum commission Enforced flag
	DefaultMinimumCommissionEnforced bool = true
)
```

Default `ante.GenesisState`
```go
// Ante params default values
var (
	// Default minimum commission rate
	DefaultMinimumCommission sdk.Dec = sdk.NewDecWithPrec(10, 2) // 10%
)
```

### CODE REVIEW

**Review by dylanschultzie — APPROVED** [review]
Nice job, looks like you covered all the cases:
1. if val_commission < minimum, upgrade it
2. don't allow val to create at under minimum
3. don't allow val to edit to below minimum
4. DO allow to edit above minimum
5. DO allow to create above minimum

The only case I see missing is creating *with* the minimum commission.

**Comment by BlockNgine** [comment]
> Nice job, looks like you covered all the cases:
> 
> 1. if val_commission < minimum, upgrade it
> 2. don't allow val to create at under minimum
> 3. don't allow val to edit to below minimum
> 4. DO allow to edit above minimum
> 5. DO allow to create above minimum
> 
> The only case I see missing is creating _with_ the minimum commission.

Just seen this PR; nice!
Think these switch case statements will handle new validator creation messages
https://github.com/terra-money/core/pull/47/files#diff-e3bd85f729e3ebf3e9b2bb00eba3d579a0e53d80cc8a4750151279c9041ee81cR39

**Review by Vritra4 — APPROVED** [review]

File: app/ante/min_commission.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+			case *stakingtypes.MsgEditValidator:
+				// if commission rate is nil, it means only
+				// other fields are affected - skip
+				if msg.CommissionRate == nil {
+					break
+				}
+
+				if msg.CommissionRate.LT(minimumCommission) {
+					return sdkerrors.Wrap(sdkerrors.ErrUnauthorized,
+						fmt.Sprintf("commission can't be lower than %s%%", minimumCommission.String()),
```
Comment (JSHan94) [inline app/ante/min_commission.go:56]: it shows `0.1%`, `%` should be removed

File: app/ante/min_commission.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+) (newCtx sdk.Context, err error) {
+	if min.anteKeeper != nil && min.anteKeeper.MinimumCommissionEnforced(ctx) {
+		minimumCommission := min.anteKeeper.GetMinimumCommission(ctx)
+
+		msgs := tx.GetMsgs()
+		validMsg := func(m sdk.Msg) error {
+			switch msg := m.(type) {
+			case *stakingtypes.MsgCreateValidator:
+				// prevent new validators joining the set with
+				// commission set below 5%
```
Comment (JSHan94) [inline app/ante/min_commission.go:40]: 10%, not 5%

**Pull request closed without merging**